message.go

```diff
@@ -513,14 +513,78 @@ type Envelope struct {
 	MessageId string
 }
 
+// Permutations of the layouts defined in RFC 5322, section 3.3.
+var dateLayouts = [...]string{
+	MessageDateTimeLayout, // popular, try it first
+	"2 Jan 2006 15:04:05 -0700",
+	"2 Jan 2006 15:04:05 MST",
+	"2 Jan 2006 15:04:05 -0700 (MST)",
+	"2 Jan 2006 15:04 -0700",
+	"2 Jan 2006 15:04 MST",
+	"2 Jan 2006 15:04 -0700 (MST)",
+	"2 Jan 06 15:04:05 -0700",
+	"2 Jan 06 15:04:05 MST",
+	"2 Jan 06 15:04:05 -0700 (MST)",
+	"2 Jan 06 15:04 -0700",
+	"2 Jan 06 15:04 MST",
+	"2 Jan 06 15:04 -0700 (MST)",
+	"02 Jan 2006 15:04:05 -0700",
+	"02 Jan 2006 15:04:05 MST",
+	"02 Jan 2006 15:04:05 -0700 (MST)",
+	"02 Jan 2006 15:04 -0700",
+	"02 Jan 2006 15:04 MST",
+	"02 Jan 2006 15:04 -0700 (MST)",
+	"02 Jan 06 15:04:05 -0700",
+	"02 Jan 06 15:04:05 MST",
+	"02 Jan 06 15:04:05 -0700 (MST)",
+	"02 Jan 06 15:04 -0700",
+	"02 Jan 06 15:04 MST",
+	"02 Jan 06 15:04 -0700 (MST)",
+	"Mon, 2 Jan 2006 15:04:05 -0700",
+	"Mon, 2 Jan 2006 15:04:05 MST",
+	"Mon, 2 Jan 2006 15:04:05 -0700 (MST)",
+	"Mon, 2 Jan 2006 15:04 -0700",
+	"Mon, 2 Jan 2006 15:04 MST",
+	"Mon, 2 Jan 2006 15:04 -0700 (MST)",
+	"Mon, 2 Jan 06 15:04:05 -0700",
+	"Mon, 2 Jan 06 15:04:05 MST",
+	"Mon, 2 Jan 06 15:04:05 -0700 (MST)",
+	"Mon, 2 Jan 06 15:04 -0700",
+	"Mon, 2 Jan 06 15:04 MST",
+	"Mon, 2 Jan 06 15:04 -0700 (MST)",
+	"Mon, 02 Jan 2006 15:04:05 MST",
+	"Mon, 02 Jan 2006 15:04:05 -0700 (MST)",
+	"Mon, 02 Jan 2006 15:04 -0700",
+	"Mon, 02 Jan 2006 15:04 MST",
+	"Mon, 02 Jan 2006 15:04 -0700 (MST)",
+	"Mon, 02 Jan 06 15:04:05 -0700",
+	"Mon, 02 Jan 06 15:04:05 MST",
+	"Mon, 02 Jan 06 15:04:05 -0700 (MST)",
+	"Mon, 02 Jan 06 15:04 -0700",
+	"Mon, 02 Jan 06 15:04 MST",
+	"Mon, 02 Jan 06 15:04 -0700 (MST)",
+}
+
+// Try parsing the date based on the layouts defined in RFC 5322, section 3.3.
+// Inspired by https://github.com/golang/go/blob/master/src/net/mail/message.go
+func (e *Envelope) ParseDate(maybeDate string) (time.Time, error) {
```

Comment:
maybe organize all the new Date/Time stuff into a new file, date.go? probably want to be able to parse all the IMAP-style date/times as well.